base/ltfssaxes.dtx

```diff
@@ -1195,59 +1208,110 @@
+%<latexrelease>\let\ulcdefault\@undefined
+%<latexrelease>\let\swshape\@undefined
+%<latexrelease>\let\swdefault\@undefined
+%<latexrelease>\let\sscshape\@undefined
+%<latexrelease>\let\sscdefault\@undefined
+%<latexrelease>
+%<latexrelease>\EndIncludeInRelease
+%<*2ekernel>
+%    \end{macrocode}
+%
 %
 % \subsection{Mapping rules for shape combinations}
 %
 %    Many of the entries are commented out as we will get that result
 %    without any entry.
 %    \begin{macrocode}
+%</2ekernel>
+%<*2ekernel|latexrelease>
+%<latexrelease>\IncludeInRelease{2025/06/01}%
+%<latexrelease>     {\DeclareFontShapeChangeRule}{Rules for ssc and sw}%
+%
 %\DeclareFontShapeChangeRule {n}{n}   {n}   {}
 \DeclareFontShapeChangeRule {n}{it}  {it}  {sl}
 \DeclareFontShapeChangeRule {n}{sl}  {sl}  {it}
-%\DeclareFontShapeChangeRule {n}{sw}  {sw}  {}
+%    \end{macrocode}
+%    If \texttt{sw} is requested but not available (not many font
+%    families offer it) we try to fallback to \texttt{it} instead of
+%    \texttt{sw}. That isn't always perfect, because some swash shapes
+%    are actually upright, but it is only a fallback and most of the
+%    time it would be better than \texttt{n}.
+% \changes{v1.0k}{2024/12/11}{Rationalize \texttt{sw} shape change
+%    rules (gh/1581)}
+%    \begin{macrocode}
+\DeclareFontShapeChangeRule {n}{sw}  {sw}  {it}
 %\DeclareFontShapeChangeRule {n}{sc}  {sc}  {}
 \DeclareFontShapeChangeRule {n}{ulc} {n}   {}
-\DeclareFontShapeChangeRule {n}{up} {n}    {}
+\DeclareFontShapeChangeRule {n}{up}  {n}   {}
+%    \end{macrocode}
+%    For the \texttt{ssc} shape we make the following general
+%    assumptions: if \texttt{ssc}\meta{X} exists then
+%    \texttt{sc}\meta{X} and \meta{X} also exist in the font.  If the
+%    \texttt{ssc}\meta{X} shape doesn't exist but the user had asked
+%    for \texttt{ssc} we try to replace it by \texttt{sc} because we
+%    then assume that the current font family simply doesn't have any
+%    \texttt{ssc} shapes.  However, if we are already in some
+%    \texttt{ssc} shape and a change is requested we know that at
+%    least some \texttt{ssc} shapes exist for the current font family,
+%    so rather than falling back to some \texttt{sc} we try to stay
+%    within \texttt{ssc} changes in a fallback situation.
+% \changes{v1.0k}{2024/12/11}{Add \texttt{ssc} shape change
+%    rules (gh/1581)}
+%    \begin{macrocode}
+\DeclareFontShapeChangeRule {n}{ssc} {ssc} {sc}
 %    \end{macrocode}
 %
 %    \begin{macrocode}
 %\DeclareFontShapeChangeRule {it}{n}   {n}     {}
 %\DeclareFontShapeChangeRule {it}{it}  {it}    {}
 \DeclareFontShapeChangeRule {it}{sl}  {sl}    {it}
-%\DeclareFontShapeChangeRule {it}{sw}  {sw}    {}
+\DeclareFontShapeChangeRule {it}{sw}  {sw}    {it}
 %    \end{macrocode}
 %    If neither \texttt{scit} nor \texttt{scsl} exist then \texttt{sc}
 %    will be used as a fallback albeit with a log entry, so except for
 %    the latter there will be no change for CM or Latin Modern fonts.
```

Comment:
I have not checked carefully all the details in the code below.

Can do this later if required! 

One observation now:
Should the alignments of the arguments be tidied up?

Reply:
they are sort of (if you remove the % on the lines that have one)

Reply:
I now went over all rules and tried to normalize/align their arguments to make them easier to compare.